cmd/kubeless/function.go

```diff
@@ -100,14 +119,246 @@ func parseMemory(mem string) (resource.Quantity, error) {
+}
+
+func (f *defaultPortForwarder) ForwardPorts(method string, url *url.URL, opts k8scmd.PortForwardOptions) error {
+	dialer, err := remotecommand.NewExecutor(opts.Config, method, url)
+	if err != nil {
+		return err
+	}
+	fw, err := portforward.New(dialer, opts.Ports, opts.StopChannel, opts.ReadyChannel, f.cmdOut, f.cmdErr)
+	if err != nil {
+		return err
+	}
+	return fw.ForwardPorts()
+}
+
+func getLocalPort() (string, error) {
+	for i := 30000; i < 65535; i++ {
+		conn, err := net.Dial("tcp", "127.0.0.1:"+strconv.Itoa(i))
+		if err != nil {
+			return strconv.Itoa(i), nil
+		}
+		conn.Close()
+	}
+	return "", errors.New("Can not find an unassigned port")
+}
+
+const (
+	maxRetries       = 5
+	defaultTimeSleep = 1
+)
+
+func waitForHTTPServer(host, port string) error {
+	// Build the request
+	req, err := http.NewRequest(
+		"GET",
+		"http://"+host+":"+port,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Execute the request
+	resp, err := http.DefaultClient.Do(req)
+	retries := 0
+	for err != nil && retries < maxRetries {
+		retries++
+		time.Sleep(time.Duration(time.Second))
+		resp, err = http.DefaultClient.Do(req)
+	}
+	resp.Body.Close()
+	return err
+}
+
+func runPortForward(f cmdutil.Factory, k8sClient *rest.RESTClient, podName, port string) (err error) {
+	clientset, err := f.ClientSet()
+	if err != nil {
+		return err
+	}
+	k8sClientConfig, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	portSlice := []string{port + ":9000"}
+	go func() {
+		devNull, err := os.Open(os.DevNull)
+		if err != nil {
+			logrus.Fatal(err)
+		}
+		pfo := k8scmd.PortForwardOptions{
+			RESTClient: k8sClient,
+			Namespace:  "kubeless",
+			Config:     k8sClientConfig,
+			PodName:    podName,
+			PodClient:  clientset.Core(),
+			Ports:      portSlice,
+			PortForwarder: &defaultPortForwarder{
+				cmdOut: devNull,
+				cmdErr: os.Stderr,
+			},
+			StopChannel:  make(chan struct{}, 1),
+			ReadyChannel: make(chan struct{}),
+		}
+		err = pfo.RunPortForward()
+		if err != nil {
+			logrus.Fatalf("Connection failed: %v", err)
+		}
+
+	}()
+	return
+}
+
+func getMinioURLFromOutsideCluster() (url string, err error) {
+	f := cmdutil.NewFactory(nil)
+	k8sClient, err := f.RESTClient()
+	if err != nil {
+		return
+	}
+
+	pods := &v1.PodList{}
+	pod := v1.Pod{}
+	k8sClient.Get().Namespace("kubeless").Resource("pods").Do().Into(pods)
+	for i := range pods.Items {
+		if pods.Items[i].Labels["kubeless"] == "minio" {
+			pod = pods.Items[i]
+			break
+		}
+	}
+
+	port, err := getLocalPort()
+	if err != nil {
+		return
+	}
+
+	// Minio requires to serve content at "root" eg. minio.com:9000
+	// so instead of using the Kubernetes Proxy we need to run
+	// a port forward to Minio's pod
+	runPortForward(f, k8sClient, pod.Name, port)
+	err = waitForHTTPServer("localhost", port)
+	if err != nil {
+		return
+	}
+	url = "localhost:" + port
+	return
+}
+
+func getMinioObject(minioClient minio.Client, bucket, name string) (object []byte, err error) {
```

Comment:
This should return an `io.Reader`, rather than reading the entire file into memory.